message_ix/macro.py

```diff
@@ -1,61 +1,399 @@
+        val = ((gdp_calibrate.loc[dict['baseyearmacro'], row.name] / 1000) **
+               rho.loc[
+                   row.name] - partmp.sum(axis=0) / (
+                           (k0.loc[row.name] / 1000) ** (
+                           rho.loc[row.name] * dict['kgdp'].loc[row.name])))
+        return val
+
+    aconst = partmp.apply(calc_aconst, axis=1)
+
+    # ------------------------------------------------------------------------
+    # add sets needed for MACRO (if they don't exist)
+    # ------------------------------------------------------------------------
+    # sets that are standard part of a MESSAGE scenario, but needed for
+    # running MACRO standalone
+    if not new.has_set("node"):
+        new.init_set("node")
+    for node in dict['region_list']:
+        new.add_set("node", node)
+
+    if not new.has_set("type_node"):
+        new.init_set("type_node")
+    new.add_set("type_node", "economy")
+
+    if not new.has_set("cat_node"):
+        new.init_set("cat_node", ["type_node", "node"])
+    for node in dict['region_list']:
+        new.add_set("cat_node", f"economy.{node}")
+
+    if not new.has_set("year"):
+        new.init_set("year")
+    for year in dict['period_list']:
+        new.add_set("year", str(year))
+
+    if not new.has_set("commodity"):
+        new.init_set("commodity")
+    for commodity in dict['commodity_list']:
+        new.add_set("commodity", commodity)
+
+    if not new.has_set("level"):
+        new.init_set("level")
+    new.add_set("level", level)
+
+    # set of MACRO base year
+    if not new.has_set("type_year"):
+        new.init_set("type_year")
+    new.add_set("type_year", "baseyear_macro")
+    new.add_set("type_year", "initializeyear_macro")
+
+    if not new.has_set("cat_year"):
+        new.init_set("cat_year", ["type_year", "year"])
+    new.add_set("cat_year", f"baseyear_macro.{dict['baseyearmacro']}")
+    new.add_set("cat_year", f"initializeyear_macro.{dict['baseyearmacro']}")
+
+    # set of MACRO sectors
+    if not new.has_set("sector"):
+        new.init_set("sector")
+    for sector in dict['sector_list']:
+        new.add_set("sector", sector)
+
+    # mapping of MACRO sectors to MESSAGE commodity/level combinations
+    if not new.has_set("mapping_macro_sector"):
+        new.init_set("mapping_macro_sector", ["sector", "commodity", "level"])
+    for k, v in sector_commodity_mapping.items():
+        new.add_set("mapping_macro_sector", [k, v, level])
+
+    # only needed for setting up a standalone MACRO model in a MESSAGE scheme
+    # new.add_set("technology","dummy")
+
+    # ------------------------------------------------------------------------
+    # add parameters needed for MACRO
+
+    # storage of MESSAGE results in GDX for standalone use of MACRO
+    # (including calibration process)
+
+    # demand
+    if not new.has_par("demand_MESSAGE"):
+        new.init_par("demand_MESSAGE", ["node", "sector", "year"])
+    for node in dict['region_list']:
```

Comment:
it should be possible to pass over the entire dataframe as is

Reply:
Changed to former way proposed in first MACRO.py version with iteration over sets.